Mathlib/Computability/NFA.lean

```diff
@@ -115,6 +122,210 @@ theorem pumping_lemma [Fintype σ] {x : List α} (hx : x ∈ M.accepts)
   rw [← toDFA_correct] at hx ⊢
   exact M.toDFA.pumping_lemma hx hlen
 
+section Closure
+
+/-!
+### NFA Closure Properties
+
+Constructions of NFAs corresponding to operations on languages.
+-/
+
+section Reversal
+
+section Auxilary
+
+/-- `M.unstep` is the reverse of `M.step`. -/
+def unstep (s : σ) (a : α) : Set σ := {s' | s ∈ M.step s' a}
+
+theorem mem_unstep {s t : σ} {a : α} : s ∈ M.unstep t a ↔ t ∈ M.step s a := Iff.rfl
+
+/-- Reversed analog of `M.stepSet S a`:
+  `M.unstepSet S a` is the union of `unstep M s a` for all `s ∈ S`.
+  It computes the set of states that have a transition in `M`
+  to a state in `S`. -/
+def unstepSet (S : Set σ) (a : α) : Set σ := ⋃ s ∈ S, M.unstep s a
+
+theorem mem_unstepSet {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, s ∈ M.unstep t a := by
+  simp [unstepSet]
+
+theorem mem_unstepSet_step {s : σ} {S : Set σ} {a : α} :
+    s ∈ M.unstepSet S a ↔ ∃ t ∈ S, t ∈ M.step s a := by
+  simp [mem_unstepSet, mem_unstep]
+
+/-- Reversed analog of `M.evalFrom S x`:
+  `M.rewindFrom S x` computes all possible reversed paths through `M` with
+  input `x` starting at an element of `S`.
+  Even though `M.rewindFrom S x` voyages across `M` in reverse, `x` is processed
+  forward, from left to right. -/
```

Comment:
I gave some thought into how to properly convey this, what do you think about this formulation?

```suggestion
  `M.rewindFrom S x` is the set of all starting points such that the automaton could be in one of the states
  in `S` after consuming `x.reverse`. -/
```

Reply:
So this is definitely true, as proven in `rewindFrom_iff_evalFrom_reverse`, but I think that the docstring for `rewindFrom` should still say exactly what's doing. It's not directly reversing the input string `x`, but going backwards through the transitions via `unstep`.
My comment could probably be better phrased, but it should probably convey that:
- `x` is consumed in order, not-reversed, from left to right,
- It goes through the transitions of `M` backwards, starting from states in `S`